### PR TITLE
feat: declarative listbox

### PR DESCRIPTION
**Pull request by gkatsev**
This defines `media-chrome-listbox`, with a role of `listbox`, and `media-chrome-listitem` with a role of `option`. These internally are basically a `ul` and `li`.
It uses element focus and "roving tabindex" for focus and selection.
`tabindex="0"` on the `media-chrome-listitem` means focused/active. `aria-selected="true"` means that the item is selected. Multi-select is supported by adding `aria-multiselectable="true"` on the `media-chrome-listbox`, which won't remove `aria-selected` from other options before applying for the current option.

TODO:
- [x] ignore disabled items
- [x] typeahead should start at current item
- [x] move example out of advanced.html
- [x] cache assignedElements/slot

### CODE REVIEW

File: examples/advanced.html
Comment (gkatsev) [inline examples/advanced.html:41]: this example can probably move into a separate file.

File: src/js/media-chrome-listbox.js
Comment (luwes) [inline src/js/media-chrome-listbox.js:175]: nit (non-blocking): is the empty newline not too spacey?

File: src/js/media-chrome-listbox.js
Comment (luwes) [inline src/js/media-chrome-listbox.js:200]: could this be undefined? might need `...)?.[0]`

File: src/js/media-chrome-listbox.js
Comment (luwes) [inline src/js/media-chrome-listbox.js:42]: nit (nb): might wanna cache this query, `this.shadowRoot.querySelector('slot')` is used 8 times in this class

File: src/js/media-chrome-listbox.js
Comment (luwes) [inline src/js/media-chrome-listbox.js:49]: B: don't forget console.log();

File: src/js/media-chrome-listbox.js
Comment (luwes) [inline src/js/media-chrome-listbox.js:233]: nit (nb): newline

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:175]: I think better to have extra whitespace than not enough :)
But I'll keep it in mind as this gets closer to completion.

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:200]: Theoretically possible, but I think in practice unlikely.
However, yeah, we should still probably do a null check.

**Comment by gkatsev** [comment]
this is good to review. One thing I'd like to re-implement https://github.com/muxinc/media-chrome/pull/282 on top of this PR before merging this PR in case there's things that come up there that might be helpful.

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [inline src/js/media-chrome-listbox.js:254]: question(blocking): Does this mean it's invalid to search for, e.g. "oo"? Should it be (honest question)?

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [inline src/js/media-chrome-listbox.js:272]: idea(non-blocking): Should we make this an attr/prop with 500ms the default to help w/a11y for folks who need more time typing?

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
Will do a second pass on this with some smoke testing permutations, but overall 😎 (a couple of small asks).

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [inline src/js/media-chrome-listbox.js:69]: issue(blocking): We can end up with weird edge cases here if `disabled` is set dynamically, especially if the item was also `selected`. Should we consider adding a well-defined event for `disabled`? We could rely on mutation observer(s)? Thoughts

Comment (gkatsev) [reply src/js/media-chrome-listbox.js:69]: So, my thought is that `disabled` should win over "selected". If a user is adding `disabled` they probably should also unselect the item as well. We currently don't have a `selectedElement` getter (but we should add one for #282) but I think it would just return nothing if the selected element ends up disabled.

I'm not quite following why we'd need a mutation observer or a disabled event. Are you saying we should deselect `disabled` items for the user?

Comment (cjpillsbury) [reply src/js/media-chrome-listbox.js:69]: > Are you saying we should deselect disabled items for the user?

Yes, sorry that was my assumption and also wasn't made clear. If we aren't in a multiselect mode, we can get into weird states if we disable a selected and then re-enable (...undisable?) after having selected another item.

Comment (gkatsev) [reply src/js/media-chrome-listbox.js:69]: Interesting, so, with a `<select>` element, when an option gets disabled, it's still set as the selected item on the select. You just can't re-select it.

So, I think the fix here is that when we reset selected, we don't ignore disabled items. We only ignore selected items for movement.

Comment (gkatsev) [reply src/js/media-chrome-listbox.js:69]: Looks like I already thought of that and forgot to point it out, haha
https://github.com/muxinc/media-chrome/pull/365/files#diff-2b6b4ea5a2bc8a7b804b26adcccf0cc24948c120c635ca3f64e7929241cf4413R178

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:254]: yes, with the current implementation, it isn't possible. Off the top of my head, not sure how to support both.

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:272]: yeah, I can see the benefit of having it.

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [inline src/js/media-chrome-listbox.js:25]: suggestion(non-blocking): If this is likely to be a base class for the dynamic/data driven implementation(s), I suspect some of these likely should not be private (and ECMA doesn't have protected access modifiers 👎)

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [reply src/js/media-chrome-listbox.js:254]: It looks like the reference implementation for listbox doesn't treat repeated keys as a special case:
https://codepen.io/pen?editors=0010
(referenced here: https://www.w3.org/WAI/ARIA/apg/example-index/listbox/listbox-scrollable.html)

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [inline src/js/media-chrome-listbox.js:259]: the reference impl handles this via the delay (relates to above convo).

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:25]: possibly, though, might be useful to have actual getters for things. Like, we probably need a `selectedItem` and `selectedItems` getters and maybe an `options` getter too.

Originally, I was thinking this should stay in PR until we build #282 on top of this, but maybe we can merge this if it's "good enough" but not expose it from `index.js` until that time. That way, we don't have a long-lasting PR but we also don't give folks a half baked listbox unless they explicitly ask for it.

File: src/js/media-chrome-listbox.js
Comment (heff) [inline src/js/media-chrome-listbox.js:182]: This seems to assume that you can turn off a selection and have no item selected. Thinking of something like the  playback rate menu, you can't turn off playback rate. Something always has to be selected. Should we be making that an option at this level, or should whatever's using this handle that?

File: src/js/media-chrome-listbox.js
Comment (heff) [reply src/js/media-chrome-listbox.js:236]: Whitespace warrior Wes. :)

I don't know why, but anything sitting right on top of an if() feels wrong to me. Gary and I have clearly worked in the same codebases. :)

File: src/js/media-chrome-listitem.js
Comment (heff) [inline src/js/media-chrome-listitem.js:98]: Does it use these? Not totally familiar with how keysUsed is used, but this feels handled by listbox.

**Review by heff — APPROVED** [review]
Good stuff 👍 None of my comments are blocking. I approve from what I see, assuming other threads are answered.

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:182]: I think that would happen at a higher level, like the playback rate menu, rather than here. Because it's reasonable to be able to unselect things for a generic listbox.

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:236]: > Gary and I have clearly worked in the same codebases. :)

Yeah, we might've 😆

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:272]: Given that the w3c example implementation uses 500, I'd say we leave it as is until we need it.

File: src/js/media-chrome-listbox.js
Comment (gkatsev) [reply src/js/media-chrome-listbox.js:254]: Let me see if I can get both working. I personally find the repeated keys thing super useful. If it's not a lot of code and doesn't take too long to implement, I'll do it, otherwise I'll take it out.

Comment (gkatsev) [reply src/js/media-chrome-listbox.js:254]: I played around with how Windows Explorer behaves in this case, and I think I have a way forward here.
Basically, you'd still want to accumulate the keys, but if it's the same letter, just treat it as a single key unless there's a better match. Shouldn't be too bad to add in.

Comment (gkatsev) [reply src/js/media-chrome-listbox.js]: Pushed a change for this

File: src/js/media-chrome-listbox.js
Comment (cjpillsbury) [reply src/js/media-chrome-listbox.js:260]: Still have some concerns around the "duplicate key" approach here, but I don't think it needs to be a blocker since followups can be isolated if we decide to change this behavior.

**Review by cjpillsbury — APPROVED** [review]
🚢 🇮🇹

**Pull request merged**
